### PR TITLE
IF: Test: Transition to instant-finality with multiple producers

### LINKED ISSUE #1512: IF: Add integration test for transition from legacy consensus to IF consensus
Depends on https://github.com/AntelopeIO/leap/issues/1522 and https://github.com/AntelopeIO/leap/issues/1548.

Many of the tests transition from pre-IF to IF to carry out the tests (see https://github.com/AntelopeIO/leap/issues/1510). But they transition with 1 BP.

This integration test should test the transition with multiple BPs. To keep test runtime short, stick with 4 BPs.

### PR DESCRIPTION
**Pull request by heifner**
- Adds integration test that verifies transition to instant-finality works with multiple producers producing before transition
- Adds integration test for `setprods` after transition to instant-finality
- Fix issue found with `setprods` to activate new proposer policy at correct block (the first block of the producer)

Resolves #1512

### CODE REVIEW

File: tests/TestHarness/Cluster.py [hunk 1]
```diff
@@ -993,34 +995,33 @@ def parseClusterKeys(totalNodes):
...
-        # pnodes does not include biosNode
-        if Utils.Debug: Utils.Print(f"threshold: {threshold}, numFins: {numFins}, pnodes: {pnodes}")
+        if Utils.Debug: Utils.Print(f"threshold: {threshold}, numFins: {numFins}")
         setFinStr =  f'{{"finalizer_policy": {{'
         setFinStr += f'  "threshold": {threshold}, '
         setFinStr += f'  "finalizers": ['
         finNum = 1
-        for n in launcher.network.nodes.values():
-            if n.index == Node.biosNodeId and not biosFinalizer:
+        for n in (self.nodes + [self.biosNode]):
```
Comment (linh2931) [inline tests/TestHarness/Cluster.py:1019]: Can this for loop be combined with the one above? They are in similar shape.

File: tests/TestHarness/Cluster.py [hunk 2]
```diff
@@ -1457,6 +1461,30 @@ def cleanup(self):
             os.remove(f)
 
     # Create accounts, if account does not already exist, and validates that the last transaction is received on root node
```
Comment (linh2931) [inline tests/TestHarness/Cluster.py:1463]: This comment should be moved down. Add a new comment about setProds if needed.

File: tests/transition_to_if.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+try:
+    TestHelper.printSystemInfo("BEGIN")
+
+    cluster.setWalletMgr(walletMgr)
+
+    Print(f'producing nodes: {pnodes}, topology: {topo}, delay between nodes launch: {delay} second{"s" if delay != 1 else ""}')
+
+    Print("Stand up cluster")
+    if args.plugin:
```
Comment (linh2931) [inline tests/transition_to_if.py:44]: Will we ever test with extra plugins? If so, add the plugins to the test command in CMakeLists.txt.

File: tests/TestHarness/Cluster.py [hunk 4]
```diff
@@ -1044,6 +1045,7 @@ def activateInstantFinality(self, launcher, biosFinalizer, pnodes):
         if not self.biosNode.waitForTransFinalization(transId, timeout=21*12*3):
```
Comment (linh2931) [inline tests/TestHarness/Cluster.py:1045]: So is the host function sent to the biosNode first even if biosFinalizer is false?

File: tests/TestHarness/Cluster.py (hunk 1 above)
Comment (heifner) [reply tests/TestHarness/Cluster.py:1019]: I don't think so, need to have `threshold` calculated first.

File: tests/TestHarness/Cluster.py (hunk 4 above)
Comment (heifner) [reply tests/TestHarness/Cluster.py:1045]: Yes. Doesn't matter which node you look for the transaction on. Maybe we should rename this python method now. It has been there for a long time. It means is the transaction in a LIB block.

**Review by linh2931 — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: TEST
summary: Integration test that verifies transition to instant-finality works with multiple producers.
Note:end

**Pull request merged**